File: apps/visualizer/backend/api/images/frame_substance.py
"""Frame substance per-image read and mutation routes."""

from __future__ import annotations

from flask import jsonify
from spectree import Response
from utils.db import with_db
from utils.lr_catalog_write import (
    describe_lr_catalog_write_status,
    read_cull_keyword_present,
    remove_cull_keyword,
    write_cull_keyword,
)
from utils.responses import error_bad_request, error_not_found, error_server_error

from api.openapi import spec
from api.schemas.frame_substance import (
    CullKeywordMutationResponse,
    FrameSubstanceInstrument,
    FrameSubstanceOverrideResponse,
    FrameSubstanceResponse,
)
from api.schemas.jobs import ErrorBody
from lightroom_tagger.core.database import (
    delete_frame_substance_override,
    get_frame_substance_verdict,
    get_image,
    get_latest_finished_frame_substance_run,
    has_excusal_channel_hint,
    has_frame_substance_override,
    insert_frame_substance_override,
    is_frame_substance_flagged,
    is_frame_substance_verdict_stale,
    library_write,
)

from .catalog import catalog_bp


def _build_frame_substance_response(db, image_key: str) -> dict:
    verdict_row = get_frame_substance_verdict(db, image_key)
    has_run = get_latest_finished_frame_substance_run(db) is not None
    catalog_status = describe_lr_catalog_write_status()

    instrument: FrameSubstanceInstrument | None = None
    restore_tier: str | None = None
    verdict_value = None
    unknown_reason = None
    detector_version = None
    judged_at = None

    if verdict_row is not None:
        verdict_value = str(verdict_row.get("verdict") or "")
        unknown_reason = str(verdict_row.get("unknown_reason") or "") or None
        detector_version = str(verdict_row.get("detector_version") or "") or None
        judged_at = str(verdict_row.get("judged_at") or "") or None
        if verdict_value in ("void", "illegible"):
            tier = "A" if verdict_value == "void" else "B"
            instrument = FrameSubstanceInstrument(
                kind="pixel_detector",
                verdict=verdict_value,
                tier=tier,
                advisory=False,
            )
            restore_tier = tier
    if instrument is None and has_excusal_channel_hint(db, image_key):
        instrument = FrameSubstanceInstrument(
            kind="excusal_channel",
            verdict=None,
            tier=None,
            advisory=True,
        )

    has_cull_keyword = None
    if catalog_status.available:
        has_cull_keyword = read_cull_keyword_present(image_key)

    return FrameSubstanceResponse(
        image_key=image_key,
        has_detection_run=has_run,
        verdict=verdict_value,
        unknown_reason=unknown_reason,
        detector_version=detector_version,
        judged_at=judged_at,
        is_stale=is_frame_substance_verdict_stale(
            db, image_key, verdict_row=verdict_row
        ),
        has_override=has_frame_substance_override(db, image_key),
        flagged=is_frame_substance_flagged(db, image_key),
        has_cull_keyword=has_cull_keyword,
        instrument=instrument,
        restore_tier=restore_tier,
        catalog_write_available=catalog_status.available,
        catalog_write_unavailable_reason=catalog_status.reason,
    ).model_dump()


@catalog_bp.route("/<path:image_key>/frame-substance", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=FrameSubstanceResponse, HTTP_404=ErrorBody),
    tags=['images-catalog'],
)
def get_catalog_frame_substance(db, image_key: str):
    try:
        if not get_image(db, image_key):
            return error_not_found("image")
        return jsonify(_build_frame_substance_response(db, image_key))
    except Exception as e:
        return error_server_error(str(e))


@catalog_bp.route("/<path:image_key>/frame-substance/override", methods=["POST"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=FrameSubstanceOverrideResponse, HTTP_404=ErrorBody),
    tags=['images-catalog'],
)
def post_catalog_frame_substance_override(db, image_key: str):
    try:
        if not get_image(db, image_key):
            return error_not_found("image")
        with library_write(db):
            insert_frame_substance_override(db, image_key)
        return jsonify({"image_key": image_key, "has_override": True})
    except Exception as e:
        return error_server_error(str(e))


@catalog_bp.route("/<path:image_key>/frame-substance/override", methods=["DELETE"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=FrameSubstanceOverrideResponse, HTTP_404=ErrorBody),
    tags=['images-catalog'],
)
def delete_catalog_frame_substance_override(db, image_key: str):
    try:
        if not get_image(db, image_key):
            return error_not_found("image")
        with library_write(db):
            delete_frame_substance_override(db, image_key)
        return jsonify({"image_key": image_key, "has_override": False})
    except Exception as e:
        return error_server_error(str(e))


@catalog_bp.route("/<path:image_key>/cull-keyword", methods=["POST"])
@with_db
@spec.validate(
    resp=Response(
        HTTP_200=CullKeywordMutationResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-catalog'],
)
def post_catalog_cull_keyword(db, image_key: str):
    try:
        if not get_image(db, image_key):
            return error_not_found("image")
        try:
            result = write_cull_keyword(image_key)
        except RuntimeError as exc:
            return error_bad_request(str(exc))
        return jsonify({"image_key": image_key, "result": result})
    except Exception as e:
        return error_server_error(str(e))


@catalog_bp.route("/<path:image_key>/cull-keyword", methods=["DELETE"])
@with_db
@spec.validate(
    resp=Response(
        HTTP_200=CullKeywordMutationResponse,
        HTTP_400=ErrorBody,
        HTTP_404=ErrorBody,
    ),
    tags=['images-catalog'],
)
def delete_catalog_cull_keyword(db, image_key: str):
    try:
        if not get_image(db, image_key):
            return error_not_found("image")
        try:
            result = remove_cull_keyword(image_key)
        except RuntimeError as exc:
            return error_bad_request(str(exc))
        return jsonify({"image_key": image_key, "result": result})
    except Exception as e:
        return error_server_error(str(e))

File: apps/visualizer/backend-ts/src/api/images/frame-substance.ts
/**
 * Frame substance per-image read and mutation routes.
 * Port of `api/images/frame_substance.py`.
 *
 * Registered ONTO the catalog group rather than as a group of its own, matching
 * `@catalog_bp.route` in Python. That is not cosmetic: these paths live under
 * `/api/images/catalog/{image_key}/…`, and a second route group would need its own
 * `use('/images/catalog/*', …)`. Hono's `app.route()` flattens a child's middleware
 * into the parent, so both registrations would then match every catalog request and
 * open two `library.db` connections for it.
 *
 * Kept in its own module for the reason it was one in Python: the cull-keyword
 * routes write to a live Lightroom `.lrcat`, and that is the one blast radius in
 * this backend worth keeping visible.
 */
import { createRoute, z, type OpenAPIHono } from '@hono/zod-openapi';
import { ERROR_IMAGE_NOT_FOUND } from '../../constants/errors.js';
import { getImage } from '../../db/library/catalog.js';
import {
  deleteFrameSubstanceOverride,
  getFrameSubstanceVerdict,
  getLatestFinishedFrameSubstanceRun,
  hasExcusalChannelHint,
  hasFrameSubstanceOverride,
  insertFrameSubstanceOverride,
  isFrameSubstanceFlagged,
  isFrameSubstanceVerdictStale,
} from '../../db/library/frame-substance.js';
import type { LibraryEnv } from '../../db/library/with-db.js';
import { libraryWrite } from '../../db/library/write.js';
import type { Db } from '../../db/connection.js';
import {
  describeLrCatalogWriteStatus,
  readCullKeywordPresent,
  removeCullKeyword,
  writeCullKeyword,
} from '../../utils/lr-catalog-write.js';
import { jsonBody, withValidationError } from '../route-helpers.js';
import { ErrorBody } from '../schemas/errors.js';
import {
  CullKeywordMutationResponse,
  FrameSubstanceOverrideResponse,
  FrameSubstanceResponse,
} from '../schemas/frame-substance.js';

const imageKeyParams = z.object({ image_key: z.string() });

/**
 * Assemble the verdict panel payload.
 *
 * Two signals, deliberately not merged: the pixel detector's verdict (tier A for
 * `void`, tier B for `illegible`) and, only when the detector has nothing to say,
 * the advisory excusal-channel hint. Collapsing them would let a weak signal look
 * like a detector judgement in the UI.
 */
function buildFrameSubstanceResponse(
  db: Db,
  imageKey: string,
): z.infer<typeof FrameSubstanceResponse> {
  const verdictRow = getFrameSubstanceVerdict(db, imageKey);
  const hasRun = getLatestFinishedFrameSubstanceRun(db) !== null;
  const catalogStatus = describeLrCatalogWriteStatus();

  let instrument: z.infer<typeof FrameSubstanceResponse>['instrument'] = null;
  let restoreTier: 'A' | 'B' | null = null;
  let verdictValue: 'void' | 'illegible' | 'ok' | 'unknown' | null = null;
  let unknownReason: string | null = null;
  let detectorVersion: string | null = null;
  let judgedAt: string | null = null;

  if (verdictRow !== null) {
    verdictValue = verdictRow.verdict;
    // Empty strings become null, matching Python's `str(...) or None`.
    unknownReason = String(verdictRow.unknown_reason || '') || null;
    detectorVersion = String(verdictRow.detector_version || '') || null;
    judgedAt = String(verdictRow.judged_at || '') || null;
    if (verdictValue === 'void' || verdictValue === 'illegible') {
      const tier = verdictValue === 'void' ? 'A' : 'B';
      instrument = { kind: 'pixel_detector', verdict: verdictValue, tier, advisory: false };
      restoreTier = tier;
    }
  }
  if (instrument === null && hasExcusalChannelHint(db, imageKey)) {
    instrument = { kind: 'excusal_channel', verdict: null, tier: null, advisory: true };
  }

  // `null` rather than `false` when the catalog cannot be read: "we do not know" is
  // a different state from "the keyword is absent", and the UI renders them
  // differently.
  const hasCullKeyword = catalogStatus.available ? readCullKeywordPresent(imageKey) : null;

  return {
    image_key: imageKey,
    has_detection_run: hasRun,
    verdict: verdictValue,
    unknown_reason: unknownReason,
    detector_version: detectorVersion,
    judged_at: judgedAt,
    is_stale: isFrameSubstanceVerdictStale(db, imageKey, { verdictRow }),
    has_override: hasFrameSubstanceOverride(db, imageKey),
    flagged: isFrameSubstanceFlagged(db, imageKey),
    has_cull_keyword: hasCullKeyword,
    instrument,
    restore_tier: restoreTier,
    catalog_write_available: catalogStatus.available,
    catalog_write_unavailable_reason: catalogStatus.reason,
  };
}

/**
 * Add the frame-substance routes to the catalog group.
 *
 * Takes the app rather than creating one so it shares the caller's `library.db`
 * middleware; the caller must open the connection writable for POST and DELETE.
 */
export function registerFrameSubstanceRoutes(app: OpenAPIHono<LibraryEnv>): void {
  // --- read -------------------------------------------------------------------

  const getRoute = createRoute({
    method: 'get',
    path: '/images/catalog/{image_key}/frame-substance',
    tags: ['images-catalog'],
    request: { params: imageKeyParams },
    responses: withValidationError({
      200: { description: 'OK', content: jsonBody(FrameSubstanceResponse) },
      404: { description: 'Not Found', content: jsonBody(ErrorBody) },
    }),
  });

  app.openapi(getRoute, (c) => {
    const db = c.get('libraryDb');
    const imageKey = c.req.param('image_key');
    if (!getImage(db, imageKey)) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);
    return c.json(buildFrameSubstanceResponse(db, imageKey), 200);
  });

  // --- override ---------------------------------------------------------------

  const overrideResponses = withValidationError({
    200: { description: 'OK', content: jsonBody(FrameSubstanceOverrideResponse) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  });

  const postOverrideRoute = createRoute({
    method: 'post',
    path: '/images/catalog/{image_key}/frame-substance/override',
    tags: ['images-catalog'],
    request: { params: imageKeyParams },
    responses: overrideResponses,
  });

  app.openapi(postOverrideRoute, (c) => {
    const db = c.get('libraryDb');
    const imageKey = c.req.param('image_key');
    if (!getImage(db, imageKey)) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);

    libraryWrite(db, () => insertFrameSubstanceOverride(db, imageKey));
    // Reports the intended state, not a re-read; the insert is idempotent.
    return c.json({ image_key: imageKey, has_override: true }, 200);
  });

  const deleteOverrideRoute = createRoute({
    method: 'delete',
    path: '/images/catalog/{image_key}/frame-substance/override',
    tags: ['images-catalog'],
    request: { params: imageKeyParams },
    responses: overrideResponses,
  });

  app.openapi(deleteOverrideRoute, (c) => {
    const db = c.get('libraryDb');
    const imageKey = c.req.param('image_key');
    if (!getImage(db, imageKey)) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);

    libraryWrite(db, () => deleteFrameSubstanceOverride(db, imageKey));
    // Deleting an override that was never there is a successful no-op.
    return c.json({ image_key: imageKey, has_override: false }, 200);
  });

  // --- cull keyword (writes the live Lightroom catalog) -----------------------

  const cullResponses = withValidationError({
    200: { description: 'OK', content: jsonBody(CullKeywordMutationResponse) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  });

  const postCullRoute = createRoute({
    method: 'post',
    path: '/images/catalog/{image_key}/cull-keyword',
    tags: ['images-catalog'],
    request: { params: imageKeyParams },
    responses: cullResponses,
  });

  app.openapi(postCullRoute, (c) => {
    const db = c.get('libraryDb');
    const imageKey = c.req.param('image_key');
    if (!getImage(db, imageKey)) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);

    try {
      return c.json({ image_key: imageKey, result: writeCullKeyword(imageKey) }, 200);
    } catch (e) {
      // An unavailable or locked catalog is a 400 the user can act on ("close
      // Lightroom"), not a 500. Anything else is genuinely unexpected.
      if (e instanceof RangeError || (e instanceof Error && /Close Lightroom/.test(e.message))) {
        return c.json({ error: e.message }, 400);
      }
      throw e;
    }
  });

  const deleteCullRoute = createRoute({
    method: 'delete',
    path: '/images/catalog/{image_key}/cull-keyword',
    tags: ['images-catalog'],
    request: { params: imageKeyParams },
    responses: cullResponses,
  });

  app.openapi(deleteCullRoute, (c) => {
    const db = c.get('libraryDb');
    const imageKey = c.req.param('image_key');
    if (!getImage(db, imageKey)) return c.json({ error: ERROR_IMAGE_NOT_FOUND }, 404);

    try {
      return c.json({ image_key: imageKey, result: removeCullKeyword(imageKey) }, 200);
    } catch (e) {
      if (e instanceof RangeError || (e instanceof Error && /Close Lightroom/.test(e.message))) {
        return c.json({ error: e.message }, 400);
      }
      throw e;
    }
  });
}
